### PR TITLE
fix normalizing in different `ParamEnv`s with the same `InferCtxt`

### PR DESCRIPTION
**Pull request by lukas-code**
This PR changes the key of the projection cache from just `AliasTy` to `(AliasTy, ParamEnv)` to allow normalizing in different `ParamEnv`s without resetting caches. Previously, normalizing the same alias in different param envs would always reuse the cached result from the first normalization, which is incorrect if the projection clauses in the param env have changed.

Fixing this bug allows us to get rid of `InferCtxt::clear_caches`, which was only used by the `AutoTraitFinder`, because it requires normalizing in different param envs.

r? @fmease

### CODE REVIEW

File: compiler/rustc_infer/src/traits/project.rs [hunk 1]
```diff
@@ -78,11 +78,12 @@ pub struct ProjectionCacheStorage<'tcx> {
 #[derive(Copy, Clone, Debug, Hash, PartialEq, Eq)]
 pub struct ProjectionCacheKey<'tcx> {
     ty: ty::AliasTy<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
```
Comment (fmease) [inline compiler/rustc_infer/src/traits/project.rs:81]: won't that lead to tons of cache misses?

Comment (lukas-code) [reply compiler/rustc_infer/src/traits/project.rs:81]: Almost all code that isn't the `AutoTraitFinder` will always use the same `ParamEnv` for a single `InferCtxt`, so I don't expect there to be many cache misses if any at all. But maybe start a perf run just to confirm? (I don't have permission)

**Comment by compiler-errors** [comment]
I thought I tried this before and the perf regression was really bad 🤔

**Review by compiler-errors — APPROVED** [review]
Awesome. Glad to see this leads to a doc perf win, too.

**Comment by compiler-errors** [comment]
Ugh. See #117131 and #117542.

Here's a minimal-ish example:

```rust
use std::future::Future;

struct Size<T>(T);

trait Foo {
    fn foo<K, V>() -> impl Future<Output = Size<impl Sized>> {
        async {
            Size(())
        }
    }
}

fn main() {}
```

We can no longer rely on the RPITIT -> opaque tys having been stashed in the projection cache, so we need to use the `normalize_param_env` *everywhere*. Not certain if that regressed behavior due to incompleteness of choosing certain predicates over others in the old trait solver.

**Comment by lukas-code** [comment]
> We can no longer rely on the RPITIT -> opaque tys having been stashed in the projection cache

It's actually the opposite happening here: `assumed_wf_types` is will attempt to normalize the RPITIT in the normal `param_env` first, so we end up with RPITIT -> RPITIT in the projection cache.
https://github.com/rust-lang/rust/blob/fecb7b43093e4fa59e2d366e88bf0085ce350f29/compiler/rustc_hir_analysis/src/check/compare_impl_item.rs#L2005

When normalizing the obligation in the `normalize_param_env` later, we pick up the RPITIT -> RPITIT from the cache, so we don't normalize RPITIT -> opaque type at all.
https://github.com/rust-lang/rust/blob/fecb7b43093e4fa59e2d366e88bf0085ce350f29/compiler/rustc_hir_analysis/src/check/compare_impl_item.rs#L2038-L2046

---

This makes me think that maybe we shouldn't normalize the RPITITs in default method implementations at all, because AFAICT there is not additional type information to be gained. Because it's impossible for a default method to specify a return type that's more concrete than the RPITIT that it's defining.

**Comment by compiler-errors** [comment]
> This makes me think that maybe we shouldn't normalize the RPITITs in default method implementations at all, because AFAICT there is not additional type information to be gained. Because it's impossible for a default method to specify a return type that's more concrete than the RPITIT that it's defining.

I'm not certain what you mean -- RPITITs can specify default method bodies that constrain the opaque:

```
trait RPITIT {
    fn test() -> impl Sized { 0 }
}
```

**Comment by compiler-errors** [comment]
Oh, do you mean just in `param_env_with_gat_bounds`? I mean, if it works then sure. It's too early in the morning for me to consider the implications of not doing so though.

**Comment by lukas-code** [comment]
> I'm not certain what you mean -- RPITITs can specify default method bodies that constrain the opaque:

Yeah but type checking will only see `impl Sized` and doesn't know that it's actually `i32`.

Opposed to this, wich actually knows the concrete type for `<i32 as RPITIT>::test`:

```rust
trait RPITIT {
    fn test() -> impl Sized;
}

trait RPITIT for i32 {
    fn test() -> i32 { 0 }
}
```

> Oh, do you mean just in param_env_with_gat_bounds?

Yes, exactly. `RevealAll` etc still have to normalize the RPITIT of course.

**Comment by lukas-code** [comment]
Alright, I found the actual problem. It was a missing `super_fold_with`.

**Comment by compiler-errors** [comment]
Oh that's incredibly embarrassing but also i'm unsurprised that normalization usually catches this. Thanks for looking into this.

@bors r+

**Pull request merged**